synapse/storage/databases/main/client_ips.py

```diff
@@ -584,15 +650,15 @@ def get_recent(txn):
                 """,
                 (since_ts, user_id),
             )
-            return txn.fetchall()
+            return cast(List[Tuple[str, str, str, int]], txn.fetchall())
 
         rows = await self.db_pool.runInteraction(
             desc="get_user_ip_and_agents", func=get_recent
         )
 
         results.update(
-            ((row["access_token"], row["ip"]), (row["user_agent"], row["last_seen"]))
-            for row in rows
+            ((access_token, ip), (user_agent, last_seen))
+            for access_token, ip, user_agent, last_seen in rows
```

Comment:
How does this work? On the left, `rows` looks like a dictionary. Ohhh... is `rows` some kind of proxy object that you can treat as both a mapping and a sequence?

Reply:
This bugfix is from #10968 - the previous code was broken because `rows` is a tuple and not a dict.